app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart

```diff
@@ -0,0 +1,126 @@
+
+  final RetryOptions retryOptions;
+
+  @override
+  Future<Body> post() async {
+    if (message.data == null) {
+      log.info('no data in message');
+      throw const BadRequestException('no data in message');
+    }
+
+    // final String data = message.data!;
+    log.fine('attempting to read message ${message.data}');
+
+    final bbv2.BatchRequest batchRequest = bbv2.BatchRequest.create();
+
+    // Merge from json only works with the integer field names.
+    batchRequest.mergeFromProto3Json(jsonDecode(message.data!) as Map<String, dynamic>);
+
+    log.info('Read the following data: ${batchRequest.toProto3Json().toString()}');
+
+    /// Retry scheduling builds upto 3 times.
+    ///
+    /// Log error message when still failing after retry. Avoid endless rescheduling
+    /// by acking the pub/sub message without throwing an exception.
+    String? unscheduledBuilds;
+    try {
+      await retryOptions.retry(
+        () async {
+          final List<bbv2.BatchRequest_Request> requestsToRetry = await _sendBatchRequest(batchRequest);
+
+          // Make a copy of the requests that are passed in as if simply access the list
+          // we make changes for all instances.
+          final List<bbv2.BatchRequest_Request> requestListCopy = [];
+          requestListCopy.addAll(requestsToRetry);
+          batchRequest.requests.clear();
+          batchRequest.requests.addAll(requestListCopy);
+
+          unscheduledBuilds = requestsToRetry.map((e) => e.scheduleBuild.builder).toString();
+          if (requestsToRetry.isNotEmpty) {
+            throw InternalServerError('Failed to schedule builds: $unscheduledBuilds.');
+          }
+        },
+        retryIf: (Exception e) => e is InternalServerError,
+      );
+    } catch (e) {
+      log.warning('Failed to schedule builds on exception: $unscheduledBuilds.');
+      return Body.forString('Failed to schedule builds: $unscheduledBuilds.');
+    }
+
+    return Body.empty;
+  }
+
+  /// Returns [List<bbv2.BatchRequest_Request>] of requests that need to be retried.
+  Future<List<bbv2.BatchRequest_Request>> _sendBatchRequest(bbv2.BatchRequest request) async {
+    log.info('Sending batch request for ${request.toProto3Json().toString()}');
+
+    // Seems to be that the create is necessary here in order to be able to assign a
+    // response here. Similar to instantiation.
+    bbv2.BatchResponse response = bbv2.BatchResponse.create();
+    try {
+      response = await buildBucketClient.batch(request);
+    } catch (e) {
+      log.severe('Exception making batch Requests.');
+      rethrow;
+    }
+
+    log.info('Made ${request.requests.length} and received ${response.responses.length}');
+    log.info('Responses: ${response.responses}');
+
+    // By default, retry everything. Then remove requests with a verified response.
+    // THese are the requests in the batch request object. Just requests.
+    final List<bbv2.BatchRequest_Request> retry = request.requests;
```

Comment:
Is the null check (from old API) not necessary? I am not sure. But to be safe, shall we keep the logic same?
```
    final List<Request> retry = request.requests ?? <Request>[];
```

Reply:
Yeah that field in the BatchRequest v2 api is not nullable.